### PR TITLE
ID-1328 system resource access policy creation on boot

### PR DESCRIPTION
**Pull request by dvoet**
Ticket: https://broadworkbench.atlassian.net/browse/ID-1328

What:

create resource access policies from config at creation tome

Why:

Terra requires some system access policies for it's services, e.g. rawls, leo, etc. To date, there have only been a couple and they have been trivial. We expect more and more complicated policies soon. Better to manage them in code.

How:

A new config stanza called `resourceAccessPolicies` is read at boot time and fed into existing code to overwrite policies. The config has the same structure as the api to create policies.

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://sdarq.dsp-appsec.broadinstitute.org/jira-ticket-risk-assesment) (requires Broad Internal network access) and attached the result to the JIRA ticket

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:86]: Are we sure we want to silently do this? Since this is on-boot, I'm wondering if it would be better to hard-fail if the users don't exist, with some descriptive message saying there's a configuration error.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala]: Why is an error non-fatal? If Sam can't upsert the policies, then I would expect startup to fail, because Sam won't be behaving how I expect.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: Do we expect to be setting up non resource_type_admin policies with this functionality?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:86]: In a BEE, the user's probably won't exist because our services register their service account users on startup. Any new live environment will also have this issue. The accounts exist in our live environments because they were registered by the same process many moons ago.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: no current expectation, but it seemed just as easy to support it as to throw an exception

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala]: It seemed drastic to have sam not start up because of this. I suppose with k8s there would not be down time, but it would still be disruptive to have to deal with it. I was not thinking about it as Sam not behaving as expected. I was thinking that Sam does not really care what the policies are. Your call, I am happy to go either way and your team supports it.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala
Comment (tlangs) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:86]: Ah, good point about BEEs!

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala
Comment (tlangs) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala]: I think I'd rather have it be disruptive. A bee would not start up if a policy can't be upserted, and we require a successful `create-bee` action to succeed before merging. I agree Sam shouldn't care what the policies are, but if a policy defined in `reference.conf` can't be upserted, I would consider that an error state.

File: src/main/resources/sam.conf
Comment (dvoet) [inline src/main/resources/sam.conf:246]: this new config should replace [this script](https://github.com/broadinstitute/dsp-scripts/blob/master/terra-support-sam-policies/terra-support-sam-policies.py)

File: src/main/resources/sam.conf
Comment (rtitle) [inline src/main/resources/sam.conf:262]: What's an example for doing this for services (Leo, Rawls)? Wondering if our existing roles will work, or if it would be better to create some new ones for service management of resources..

Comment (dvoet) [reply src/main/resources/sam.conf:262]: my intent was to create new roles with hand picked actions. The workspace application role is probably ok for Leo but every other resource type probably needs other roles.

File: src/main/resources/sam.conf
Comment (dvoet) [reply src/main/resources/sam.conf:246]: see also https://github.com/broadinstitute/terra-helmfile/pull/5706

**Review by tlangs — APPROVED** [review]

**Review by samanehsan — APPROVED** [review]

**Pull request merged**
